lib/HeroBonus.cpp

```diff
@@ -375,16 +407,9 @@ int BonusList::valOfBonuses(const CSelector &select) const
 	BonusList ret;
 	CSelector limit = nullptr;
 	getBonuses(ret, select, limit);
-	ret.eliminateDuplicates();
 	return ret.totalValue();
 }
 
-void BonusList::eliminateDuplicates()

```

Comment:
You should put stacking code here, not to `totalValue()`
* cached BonusList`s will be preprocessed
* no additional checks will be needed where `getAllBonuses` used directly

Reply:
I think conceptually stacking is part of total value calculation, whereas getAllBonuses should really return all bonuses, not just some. Also, `IBonusBearer::valOfBonuses` uses caching as well, so I doubt it makes much of a difference performance-wise.

Reply:
You put stacking after cache request but it was before (f.e. in CBonusSystemNode::getAllBonuses).
And it should be no additional stacking processing inside BonusList. Stacking is IBonusBearer responsibility.

Reply:
Just because duplicates were eliminated as part of getAllBonuses before doesn't mean that's a good approach. And I don't see why stacking should be the responsibility of IBonusBearer - totalValue calculation is part of BonusList because all you need is a list of bonuses, and the same argument applies to stacking.